idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,965 @@
+        if kmax is None:
+            kmax = 30
+            _logger.warning(f"kmax was not set - using a default value of 30.")
+
+        if not isinstance(kmin, int):
+            raise ValueError(f"kmin must be an integer, but {kmin} is not an integer")
+        if not isinstance(kmax, int):
+            raise ValueError(f"kmax must be an integer, but {kmax} is not an integer")
+        if kmin < 1:
+            raise ValueError(f"kmin must be > 0, but {kmin} is provided.")
+        if kmax < 1:
+            raise ValueError(f"kmax must be > 0, but {kmax} is provided.")
+        if kmin >= kmax:
+            raise ValueError(f"kmin must be less than kmax, but {kmin} >= {kmax}")
+
+        k_values = range(kmin, kmax + 1)
+        inertia_values = []
+
+        np.random.seed(self._seed)
+
+        # Compute the inertia (SSE) for k clusters
+        for k in k_values:
+            kmeans = KMeans(n_clusters=k).fit(daily_data.transpose())
+            inertia_values.append(kmeans.inertia_)
+
+        # Identify the "elbow point"
+        elbow_point = KneeLocator(
+            k_values, inertia_values, curve="convex", direction="decreasing"
+        )
+        n_clusters = elbow_point.knee
+
+        if n_clusters is None:
+            raise ValueError(
+                f"Could not find elbow point for given kmin, kmax. Consider increasing the range of kmin, kmax."
+            )
+
+        print(f"Optimal # of clusters is: {n_clusters}")
+
+        if int(n_clusters) + 2 >= kmax:
+            _logger.warning(
+                f"Optimal number of clusters is close to kmax: {kmax}. Consider increasing kmax."
+            )
+
+        if plot == True:
+            plt.plot(k_values, inertia_values)
+            plt.axvline(x=n_clusters, color="red", linestyle="--", label="Elbow")
+            plt.xlabel("Number of clusters")
+            plt.ylabel("Inertia")
+            plt.title("Elbow Method")
+            plt.xlim(kmin, kmax)
+            plt.grid()
+
+        return int(n_clusters), inertia_values
+
+    def cluster_lmp_data(self, raw_data, n_clusters):
+        """
+        Clusters the given price signal in n_clusters. This method supports k-means, k-meteiod,...
+        techniques for clustering.
+
+        Args:
+            raw_data:   Columnar data for a given LMP signal
+            n_clusters: number of clusters desired for the data (representative days)
+
+        Returns:
+            lmp_data:   dict of representative day LMP data, indices are indexed
+                        by integers starting at 1 (ex: {1: {1: 4, 2: 3, 3: 5},
+                                                        2: {1: 1, 2: 7, 3: 3}})
+                                               Format: {day: {time_period: LMP}}
+            weights:    dict of weights for each representative day, indexed the
+                        same way as lmp_data      (ex: {1: 45, 2: 56})
+                                               Format: {day: weight}
+        """
```

Comment:
@dlakes94 this seems to resolve the Sphinx warnings:

```suggestion
        """
        Clusters the given price signal in n_clusters. This method supports k-means,
        k-medoids, etc., techniques for clustering.

        Args:
            raw_data: Columnar data for a given LMP signal.
            n_clusters: Number of clusters desired for the data (representative days).

        Returns:
            lmp_data: A dictionary of representative day LMP data, indices are indexed
                    by integers starting at 1. Example: ::

                        {1: {1: 4, 2: 3, 3: 5},
                        2: {1: 1, 2: 7, 3: 3}}

            weights: A dictionary of weights for each representative day, indexed the
                    same way as lmp_data. Example: ::

                        {1: 45, 2: 56}
        """
```

Reply:
Thank you! I will incorporate this.